lib/ansible/modules/cloud/docker/docker_node_facts.py

```diff
@@ -0,0 +1,119 @@
+#!/usr/bin/python
+#
+# (c) 2019 Piotr Wojciechowski <piotr@it-playground.pl>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+---
+module: docker_node_facts
+
+short_description: Retrieves facts about docker swarm node from Swarm Manager
+
+description:
+  - Retrieves facts about a docker node.
+  - Essentially returns the output of C(docker node inspect <name>)
+  - Must be executed on host running as Swarm Manager
+
+version_added: "2.8"
+
+options:
+  name:
+    description:
+      - The name of the node to inspect.
+      - When identifying an existing node name may be a name or a long or short container ID.
```

Comment:
I assume that if this is a container ID, it will identify the node where the container runs on? Maybe the description could be a bit more explicit.

Reply:
The API docs (https://docs.docker.com/engine/api/v1.24/#37-nodes) do not mention that a long or short container ID can be specified here; not even the docs for the latest API version (https://docs.docker.com/engine/api/v1.39/#operation/NodeInspect). I also cannot see why that should work from https://github.com/moby/moby/blob/master/daemon/cluster/helpers.go#L26.

Reply:
> The API docs (https://docs.docker.com/engine/api/v1.24/#37-nodes) do not mention that a long or short container ID can be specified here; not even the docs for the latest API version (https://docs.docker.com/engine/api/v1.39/#operation/NodeInspect). I also cannot see why that should work from https://github.com/moby/moby/blob/master/daemon/cluster/helpers.go#L26.

The `ID` identification is Swarm related, it is not the Docker host `ID`, it has different format. The `Description.Hostname` matches the Docker host `Name` attribute. Those are not container ID's so there is no long and truncated representation.

The name parameter is the same as in API and command line - it is either the `Description.Hostname` or the `ID` of the Swarm node:

```JSON
{
  *"ID": "24ifsmvkjbyhk",*
  "Version": {
    "Index": 373531
  },
  "CreatedAt": "2016-08-18T10:44:24.496525531Z",
  "UpdatedAt": "2017-08-09T07:09:37.632105588Z",
  "Spec": {
    "Availability": "active",
    "Name": "node-name",
    "Role": "manager",
    "Labels": {
      "foo": "bar"
    }
  },
  "Description": {
    *"Hostname": "bf3067039e47",*
    "Platform": {
      "Architecture": "x86_64",
      "OS": "linux"
    },
(... omitted ...)
}
```

The `Spec.Name` form the example above (example from documentation) does not exist in output on version 17.06 and 18.09 of Docker on node. In command line the filter `name` refers to Hostname

```
$ docker node ls -f 
id=          label=       membership=  name=        role= 
```       

You can manually set the `Spec.Name` attribute using` /update` POST method but you still can't use it to identify node for inspection.

To keep align with other `docker_*_facts `modules the name is required and it returns facts only about single swarm node. The only exception from this rule is `docker_image_facts` which returns information on all images if `name` is missing (there is inconsistecy in documentation as it says the name is required), As mentioned in #50428 I proposed in new module `docker_node` a state `list` that will return list of names of Swarm nodes that can be provided as a loop parameter for `docker_node_inspect`. I can also make the `name` parameter here optional and return information on all nodes. 

I think it is rather question on consistent approach to the name parameter across all `docker_*_facts` modules.

Reply:
(Finally: let's leave the `mode` / optional `name` parameter away and get this merged first; you can then add it in #50428 or a follow-up PR, so that you can use the new `module_utils` Python module.)

Reply:
Fixed the description

The `mode` proposal we can discuss, but if we leave it in `docker_node` in state `list` as a representation of `docker node ls` command we can return the list of nodes in different formats - just hostnames, just IDs, add availability information, etc. This will reflect what you can get from the command line and might be easier for many people to understand and use the data afterwards. I like the approach it returns full information on a specific node while listing nodes and other manipulations are in `docker_node`. This is more clear for me, but if most disagree with me, I will implement it in a way preferred by most people.